File: OrangeJuice.Server.Core/Configuration/CachingConfigurationProvider.cs
using System;
using System.Runtime.Caching;

namespace OrangeJuice.Server.Configuration
{
	public sealed class CachingConfigurationProvider : IConfigurationProvider
	{
		private readonly IConfigurationProvider _configurationProvider;
		private readonly MemoryCache _cache;

		public CachingConfigurationProvider(IConfigurationProvider configurationProvider, MemoryCache cache)
		{
			_configurationProvider = configurationProvider;
			_cache = cache;
		}

		public string GetValue(string key)
		{
			return RetrieveAndCache(key, () => _configurationProvider.GetValue(key), CreatePolicy);
		}

		private static CacheItemPolicy CreatePolicy()
		{
			return new CacheItemPolicy();
		}

		private T RetrieveAndCache<T>(string key, Func<T> valueFactory, Func<CacheItemPolicy> policyFactory)
		{
			var newValue = new Lazy<T>(valueFactory);
			var policy = policyFactory();
			var value = (Lazy<T>)_cache.AddOrGetExisting(key, newValue, policy);
			return (value ?? newValue).Value;
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Configuration/CachingConfigurationProvider.fs
namespace OrangeJuice.Server.FSharp.Configuration

open System

open OrangeJuice.Server.Cache
open OrangeJuice.Server.Configuration

type CachingConfigurationProvider(configurationProvider : IConfigurationProvider, cacheClient : ICacheClient) =
    interface IConfigurationProvider with
        member this.GetValue(key : string) : string =
            let valueFactory = new Func<string>(fun() -> configurationProvider.GetValue(key))
            cacheClient.AddOrGetExisting(key, valueFactory)
